WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentOnboardingAdaptor.swift

```diff
@@ -0,0 +1,72 @@
+/// route the output to a SwiftUI view for display, rather than directly displaying on the viewController that's passed in.
+final class CardPresentPaymentsOnboardingPresenterAdaptor: CardPresentPaymentsOnboardingPresenting {
+    private let stores: StoresManager
+
+    private let onboardingUseCase: CardPresentPaymentsOnboardingUseCase
+
+    private let readinessUseCase: CardPresentPaymentsReadinessUseCase
+
+    private let onboardingViewModel: CardPresentPaymentsOnboardingViewModel
+
+    private var readinessSubscription: AnyCancellable?
+
+    private(set) var onboardingScreenViewModelPublisher: AnyPublisher<CardPresentPaymentOnboardingPresentationEvent, Never>
+
+    private let onboardingScreenViewModelSubject: PassthroughSubject<CardPresentPaymentOnboardingPresentationEvent, Never> = PassthroughSubject()
+
+    init(stores: StoresManager = ServiceLocator.stores) {
+        self.stores = stores
+        onboardingUseCase = CardPresentPaymentsOnboardingUseCase(stores: stores)
+        readinessUseCase = CardPresentPaymentsReadinessUseCase(onboardingUseCase: onboardingUseCase, stores: stores)
+        onboardingViewModel = CardPresentPaymentsOnboardingViewModel(useCase: onboardingUseCase)
+        onboardingScreenViewModelPublisher = onboardingScreenViewModelSubject.eraseToAnyPublisher()
+    }
+
+    /// If the onboarding state is not `ready`, this will instruct downstream SwiftUI code to present the onboarding screen.
+    /// The CardPresentPaymentOnboardingView controls which message will be shown based on the view model we pass, which will change over time.
+    /// - Parameters:
+    ///   - viewController: This will be ignored, as other SwiftUI code is responsible for the display in this implementation.
+    ///   - completion: Callback when the onboarding is complete
+    func showOnboardingIfRequired(from viewController: ViewControllerPresenting,
+                                  readyToCollectPayment completion: @escaping () -> Void) {
+        guard readinessSubscription == nil else {
+            return
+        }
+
+        readinessUseCase.checkCardPaymentReadiness()
+
+        guard case .ready = readinessUseCase.readiness else {
+            return showOnboarding(readyToCollectPayment: completion)
+        }
+
+        completion()
+    }
+
+    private func showOnboarding(readyToCollectPayment completion: @escaping () -> Void) {
+        onboardingScreenViewModelSubject.send(.showOnboarding(onboardingViewModel))
+
+        readinessSubscription = readinessUseCase.$readiness
+            .subscribe(on: DispatchQueue.main)
+            .sink(receiveValue: { [weak self] readiness in
+                guard let self,
+                      case .ready = readiness else {
+                    return
+                }
+
+                onboardingScreenViewModelSubject.send(.onboardingComplete)
```

Comment:
I checked the UIKit version, so nice with the publisher!

Reply:
I played with it more in the experiment and I think we may have some issues actually using it to display the UI, so we may still need to rely on the POS feature gate requiring onboarding completion for a while.

It was really strange – the underlying viewmodel's state was updating, and the onboarding view's `body` was called and went through the switch case to choose the right screen, but it never changed what was displayed.

That's a problem for another day though!

Reply:
Could you share the steps to repro this if it's reproducible? We can log this first and look into this after the happy path 👍

Reply:
Quite a few ways, but one repro was along the lines of

1. Disable WooPayments on your store
2. Edit the POS eligibility checks to make sure you still see the button
3. Launch the app
4. Go to `Menu > POS Mode`
5. Try to connect the reader
6. Observe that the "WooPayments is disabled" onboarding step is correctly shown
7. Tap "Activate WooPayments"
8. Observe that it never changes what's shown on screen.

Breakpoints can show it going through the OnboardingView switch statement